lib/active_model/serializer/adapter/json_api/pagination_links.rb

```diff
@@ -0,0 +1,59 @@
+module ActiveModel
+  class Serializer
+    class Adapter
+      class JsonApi < Adapter
+        class PaginationLinks
+          FIRST_PAGE = 1
+
+          attr_reader :collection, :context
+
+          def initialize(collection, context)
+            @collection = collection
+            @context = context
+          end
+
+          def serializable_hash(options = {})
+            pages_from.each_with_object({}) do |(key, value), hash|
+              params = query_parameters.merge(page: { number: value, size: collection.size }).to_query
+
+              hash[key] = "#{url(options)}?#{params}"
+            end
+          end
+
+          private
+
+          def pages_from
+            return {} if collection.total_pages == FIRST_PAGE
+
+            {}.tap do |pages|
+              pages[:self] = collection.current_page
+
+              unless collection.current_page == FIRST_PAGE
+                pages[:first] = FIRST_PAGE
+                pages[:prev]  = collection.current_page - FIRST_PAGE
+              end
+
+              unless collection.current_page == collection.total_pages
+                pages[:next] = collection.current_page + FIRST_PAGE
+                pages[:last] = collection.total_pages
+              end
+            end
+          end
+
+          def url(options)
+            self_link = options.fetch(:links) {{}}
+            self_link.fetch(:self) {} ? options[:links][:self] : original_url
```

Comment:
Looking good.  I'm a little uncertain about what this method is doing.  It seems to be returning `options[:links][:self] || original_url`, but it's being called on each pagination link, and after `pages[:self]` is set, so am I misunderstanding what url this is returning?  It's just the scheme + host + port of the current page, right?

```ruby
page_numbers = {
  self: collection.current_page,
  first: 1,
  prev: collection.current_page - 1,
  next: collection.current_page + 1,
  last: collection.total_pages,
}
# # missing method calls?
# current_page_params = query_parameters.delete(:page)
# options[:links] ||= {}
# url = options[:links][:self] ||= original_url
page_numbers.each_with_object({}) {|(page_name, page_number), pagination_links|
  page_link = {  page: { number: value, size: collection.size } }
  page_params = query_parameters.merge(page_link) # fresh for each link
  query_params = page_params.to_query
  url = options.fetch(:links, {}).fetch(:self, nil) ? options[:links][:self] : original_url
  page_links[page_name] = "#{url}?#{query_params}"
  # require 'uri/http'
  # URI::HTTP.build(scheme: 'http', host: "www.example.com", path:, "/path",
  "query": "query").to_s
}
```